### PR TITLE
Dws

### PR DESCRIPTION
**Pull request by vineet-pruthi**
The update contains all SDK and OpenStack Client parts for the 'Data Warehouse Service' module and the unit tests.

### CODE REVIEW

File: otcextensions/osclient/dws/v1/snapshot.py [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+        parser.add_argument(
+            'cluster',
+            metavar='<cluster>',
+            help=_('ID or name of the cluster for which you want to '
+                   'create a snapshot.')
+        )
+        parser.add_argument(
+            'name',
+            metavar='<name>',
+            # required=True,
```
Comment (anton-sidelnikov) [inline otcextensions/osclient/dws/v1/snapshot.py]: not required?

File: otcextensions/tests/unit/osclient/dws/v1/fakes.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    def setUp(self):
+        super(TestDws, self).setUp()
+
+        self.app.client_manager.dws = mock.Mock()
+
+        self.client = self.app.client_manager.dws
+
+
+class FakeCluster(test_base.Fake):
+    """Fake one or more Nat Gateways."""
```
Comment (anton-sidelnikov) [inline otcextensions/tests/unit/osclient/dws/v1/fakes.py]: DWS?

**Review by anton-sidelnikov — CHANGES_REQUESTED** [review]

File: otcextensions/tests/unit/osclient/dws/v1/fakes.py (hunk 2 above)
Comment (vineet-pruthi) [reply otcextensions/tests/unit/osclient/dws/v1/fakes.py]: Updated to Dws.

File: otcextensions/osclient/dws/v1/snapshot.py (hunk 1 above)
Comment (vineet-pruthi) [reply otcextensions/osclient/dws/v1/snapshot.py]: Since name is a positional argument, i commented out required=True

**Review by anton-sidelnikov — DISMISSED** [review]

File: otcextensions/sdk/dws/v1/flavor.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+                raw_resource.pop("self", None)
+                # We want that URI props are available on the resource
+                raw_resource.update(uri_params)
+
+                value = cls.existing(
+                    microversion=microversion,
+                    connection=session._get_connection(),
+                    **raw_resource,
+                )
+                for detail in value.detail:
```
Comment (gtema) [inline otcextensions/sdk/dws/v1/flavor.py]: ugh how ugly it is, but what else

File: otcextensions/sdk/dws/v1/snapshot.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+    # capabilities
+    allow_create = True
+    allow_fetch = True
+    allow_delete = True
+    allow_list = True
+
+    # Properties
+    #: ID of the cluster for which snapshots are created.
+    cluster_id = resource.Body('cluster_id')
```
Comment (gtema) [inline otcextensions/sdk/dws/v1/snapshot.py:30]: would be great if you could sort all attributes . We try to keep them (at least in OpenStackSDK side) sorted alphabetically. Especially like one from above with tons of attributes which costs maintenance effort if unsorted

File: otcextensions/sdk/dws/v1/snapshot.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    #: Snapshot name.
+    name = resource.Body('name')
+    #: Snapshot description.
+    description = resource.Body('description')
+    #: Time when a snapshot starts to be created.
+    started = resource.Body('started')
```
Comment (gtema) [inline otcextensions/sdk/dws/v1/snapshot.py]: everywhere we have created_at. Are they serious? Please name it created_at

File: otcextensions/sdk/dws/v1/cluster.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+    # capabilities
+    allow_create = True
+    allow_list = True
+    allow_commit = True
+    allow_delete = True
+    allow_fetch = True
+    allow_patch = True
+
+    # Properties
+    #: Cluster name.
```
Comment (gtema) [inline otcextensions/sdk/dws/v1/cluster.py]: please sort attrs alphabetically.

File: otcextensions/sdk/dws/v1/cluster.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    name = resource.Body('name')
```
Comment (gtema) [inline otcextensions/sdk/dws/v1/cluster.py]: name and id attrs are present in all resources through resource.Resource. There is no need for them explicitly

File: otcextensions/sdk/dws/v1/cluster.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+    #: Data warehouse version.
+    version = resource.Body('version')
+    #: Node Type (DWS Flavor).
+    flavor = resource.Body('node_type')
+    #: Node Type ID (DWS Flavor ID).
+    flavor_id = resource.Body('node_type_id')
+    #: Nodes List.
+    nodes = resource.Body('nodes', type=list)
+    #: Number of nodes in a cluster. The value ranges from 3 to 32.
+    number_of_node = resource.Body('number_of_node')
```
Comment (gtema) [inline otcextensions/sdk/dws/v1/cluster.py]: at least number_of_nodeS, but I would prefer num_nodes as attr name. We should fix API names on SDK side and not repeat same mistakes

File: otcextensions/sdk/dws/v1/cluster.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+    #: number_of_free_node
+    number_of_free_node = resource.Body('number_of_free_node', type=int)
```
Comment (gtema) [inline otcextensions/sdk/dws/v1/cluster.py]: same as above

File: otcextensions/sdk/dws/v1/cluster.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+    #:  - AVAILABLE
+    #:  - UNAVAILABLE
+    #:  - CREATION FAILED
+    status = resource.Body('status')
+    #: Sub-status of clusters in the AVAILABLE state.
+    sub_status = resource.Body('sub_status')
+    #: Cluster management task.
+    task_status = resource.Body('task_status')
+    #: The key indicates an ongoing task.
+    action_progress = resource.Body('action_progress')
```
Comment (gtema) [inline otcextensions/sdk/dws/v1/cluster.py]: you should specify type=dict to visualize this is not something like is_action_running

File: otcextensions/sdk/dws/v1/cluster.py [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+    #: Last modification time of a cluster.
+    updated_at = resource.Body('updated')
+    #: Number of events.
+    recent_event = resource.Body('recent_event', type=int)
```
Comment (gtema) [inline otcextensions/sdk/dws/v1/cluster.py]: if this is number of events then num_recent_events

File: otcextensions/sdk/dws/v1/cluster.py [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+    #: List of private network IP addresses.
+    private_ip = resource.Body('private_ip', type=list)
+    #: Router ID, which is used for configuring cluster network.
+    router_id = resource.Body('vpc_id')
+    #: Network ID, which is used for configuring cluster network.
+    network_id = resource.Body('subnet_id')
+    #: ID of a security group, which is used for configuring cluster network
+    security_group_id = resource.Body('security_group_id')
+    #: Cluster maintenance window.
+    maintain_window = resource.Body('maintain_window', type=dict)
```
Comment (gtema) [inline otcextensions/sdk/dws/v1/cluster.py]: let's name it maintenance window. If API creator is not English-skilled we should not repeat that.

File: otcextensions/sdk/dws/v1/cluster.py [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+    #: Enterprise project ID.
+    #: If no enterprise project is specified for a cluster,
+    #: the default enterprise project ID 0 is used.
+    enterprise_project_id = resource.Body('enterprise_project_id')
+    #: Cluster scale-out details.
+    resize_info = resource.Body('resize_info', type=dict)
+    #: Guest Agent Version
+    guest_agent_version = resource.Body('guest_agent_version')
+    #: use_logical_cluster
+    use_logical_cluster = resource.Body('use_logical_cluster', type=bool)
```
Comment (gtema) [inline otcextensions/sdk/dws/v1/cluster.py]: I could not find this param in API docs. Is it something new? But anyway - flag attributes are named as "is_XXXX"

File: otcextensions/sdk/dws/v1/cluster.py [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+    #: logical_cluster_initialed
+    logical_cluster_initialed = resource.Body(
```
Comment (gtema) [inline otcextensions/sdk/dws/v1/cluster.py]: same as above

File: otcextensions/sdk/dws/v1/cluster.py [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+        'logical_cluster_initialed', type=bool)
+    #: logical_cluster_mode
+    logical_cluster_mode = resource.Body('logical_cluster_mode', type=bool)
```
Comment (gtema) [inline otcextensions/sdk/dws/v1/cluster.py]: same

File: otcextensions/sdk/dws/v1/flavor.py [hunk 16]
```diff
@@ -0,0 +1,214 @@
...
+    # Properties
+    #: Name of a node type
+    name = resource.Body('spec_name')
+    #: Name of a node type
+    spec_name = resource.Body('spec_name')
+    #: Node type details
+    detail = resource.Body('detail', type=list, list_type=FlavorDetail)
+
+    #: Memory
+    ram = resource.Computed('ram')
```
Comment (gtema) [inline otcextensions/sdk/dws/v1/flavor.py]: pls sort attrs

File: otcextensions/sdk/dws/v1/snapshot.py [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+    #: Time when a snapshot is complete.
+    finished = resource.Body('finished')
```
Comment (gtema) [inline otcextensions/sdk/dws/v1/snapshot.py]: that should be then updated_at

File: otcextensions/sdk/dws/v1/snapshot.py [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+    #: Snapshot size, in GB.
+    size = resource.Body('size', type=float)
```
Comment (gtema) [inline otcextensions/sdk/dws/v1/snapshot.py:36]: even docs explicitly says that, all examples are just integers. Can you please verify?

**Review by gtema — CHANGES_REQUESTED** [review]
Thanks Vineet,

there are few fundamental attribute naming issues that we need to fix first. Since this is going to have big impact across remaining code I was not looking deeper to other functions.

File: otcextensions/sdk/dws/v1/snapshot.py (hunk 4 above)
Comment (vineet-pruthi) [reply otcextensions/sdk/dws/v1/snapshot.py:30]: Sorted :)

File: otcextensions/sdk/dws/v1/cluster.py (hunk 6 above)
Comment (vineet-pruthi) [reply otcextensions/sdk/dws/v1/cluster.py]: Done.

File: otcextensions/sdk/dws/v1/cluster.py (hunk 7 above)
Comment (vineet-pruthi) [reply otcextensions/sdk/dws/v1/cluster.py]: removed attributes name/id from resource class if return body has name/id.

File: otcextensions/sdk/dws/v1/cluster.py (hunk 8 above)
Comment (vineet-pruthi) [reply otcextensions/sdk/dws/v1/cluster.py]: Attribute updated to num_nodes

File: otcextensions/sdk/dws/v1/cluster.py (hunk 9 above)
Comment (vineet-pruthi) [reply otcextensions/sdk/dws/v1/cluster.py]: Attr updated to num_free_nodes.

File: otcextensions/sdk/dws/v1/cluster.py (hunk 10 above)
Comment (vineet-pruthi) [reply otcextensions/sdk/dws/v1/cluster.py]: Sorry i forgot to mention type. Updated :)

File: otcextensions/sdk/dws/v1/cluster.py (hunk 11 above)
Comment (vineet-pruthi) [reply otcextensions/sdk/dws/v1/cluster.py]: Attrs Updated to num_recent_events.

File: otcextensions/sdk/dws/v1/cluster.py (hunk 12 above)
Comment (vineet-pruthi) [reply otcextensions/sdk/dws/v1/cluster.py]: Totally agree 👍  , Update to maintenance_window

File: otcextensions/sdk/dws/v1/cluster.py (hunk 13 above)
Comment (vineet-pruthi) [reply otcextensions/sdk/dws/v1/cluster.py]: These attributes are missing in API ref, but they are part of RESP Body. Going to create Bug for missing information in the document.
updated the attribute to "is_logical_cluster_enabled"

File: otcextensions/sdk/dws/v1/cluster.py (hunk 14 above)
Comment (vineet-pruthi) [reply otcextensions/sdk/dws/v1/cluster.py]: Updated attr to "is_logical_cluster_initialed"

File: otcextensions/sdk/dws/v1/flavor.py (hunk 16 above)
Comment (vineet-pruthi) [reply otcextensions/sdk/dws/v1/flavor.py]: Done.

File: otcextensions/sdk/dws/v1/snapshot.py (hunk 18 above)
Comment (vineet-pruthi) [reply otcextensions/sdk/dws/v1/snapshot.py:36]: Its specified as integer but resp body returns float size  :(
will also raise this as Bug

RESP BODY: {"snapshot":{"id":"XXXXXXXXX","name":"XXXXXXX","description":"","started":"2023-01-30T10:14:36","finished":"2023-01-30T10:19:03",**"size":0.403551,**"status":"AVAILABLE","type":"AUTO","cluster_id":"XXXX"}}

File: otcextensions/sdk/dws/v1/snapshot.py (hunk 17 above)
Comment (vineet-pruthi) [reply otcextensions/sdk/dws/v1/snapshot.py]: Fixed.

File: otcextensions/sdk/dws/v1/snapshot.py (hunk 5 above)
Comment (vineet-pruthi) [reply otcextensions/sdk/dws/v1/snapshot.py]: Fixed.

**Review by anton-sidelnikov — APPROVED** [review]

**Review by vladimirhasko — APPROVED** [review]

**Review by gtema — APPROVED** [review]

**Pull request merged**